libs/parallelism/algorithms/include/hpx/parallel/algorithms/unique.hpp

```diff
@@ -213,147 +682,101 @@ namespace hpx { namespace parallel { inline namespace v1 {
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
         return detail::unique<FwdIter>().call(std::forward<ExPolicy>(policy),
             first, last, std::forward<Pred>(pred), std::forward<Proj>(proj));
+#if defined(HPX_GCC_VERSION) && HPX_GCC_VERSION >= 100000
+#pragma GCC diagnostic pop
+#endif
     }
 
+    template <typename I, typename O>
+    using unique_copy_result = util::in_out_result<I, O>;
+
     /////////////////////////////////////////////////////////////////////////////
     // unique_copy
     namespace detail {
         /// \cond NOINTERNAL
 
         // sequential unique_copy with projection function
-        template <typename FwdIter, typename OutIter, typename Pred,
-            typename Proj>
-        std::pair<FwdIter, OutIter> sequential_unique_copy(FwdIter first,
-            FwdIter last, OutIter dest, Pred&& pred, Proj&& proj,
+        template <typename FwdIter, typename Sent, typename OutIter,
+            typename Pred, typename Proj>
+        unique_copy_result<FwdIter, OutIter> sequential_unique_copy(
+            FwdIter first, Sent last, OutIter dest, Pred&& pred, Proj&& proj,
             std::true_type)
         {
             if (first == last)
-                return std::make_pair(std::move(last), std::move(dest));
+                return unique_copy_result<FwdIter, OutIter>{
+                    std::move(first), std::move(dest)};
 
             FwdIter base = first;
 
             *dest++ = *first;
 
             while (++first != last)
             {
-                if (!hpx::util::invoke(pred, hpx::util::invoke(proj, *base),
-                        hpx::util::invoke(proj, *first)))
+                if (!HPX_INVOKE(pred, HPX_INVOKE(proj, *base),
+                        HPX_INVOKE(proj, *first)))
                 {
                     base = first;
                     *dest++ = *first;
                 }
             }
-            return std::make_pair(std::move(last), std::move(dest));
+            return unique_copy_result<FwdIter, OutIter>{
+                std::move(first), std::move(dest)};
         }
 
         // sequential unique_copy with projection function
-        template <typename InIter, typename OutIter, typename Pred,
-            typename Proj>
-        std::pair<InIter, OutIter> sequential_unique_copy(InIter first,
-            InIter last, OutIter dest, Pred&& pred, Proj&& proj,
-            std::false_type)
+        template <typename InIter, typename Sent, typename OutIter,
+            typename Pred, typename Proj>
+        unique_copy_result<InIter, OutIter> sequential_unique_copy(InIter first,
+            Sent last, OutIter dest, Pred&& pred, Proj&& proj, std::false_type)
         {
             if (first == last)
-                return std::make_pair(std::move(last), std::move(dest));
+                return unique_copy_result<InIter, OutIter>{
+                    std::move(first), std::move(dest)};
 
             auto base_val = *first;
```

Comment:
Please don't use `auto` for storing the iterator `value_type`'s to correctly handle proxy iterators.